specs/GetFavicon.md

```diff
@@ -0,0 +1,155 @@
+# Background
+A favicon (favorite icon) is a tiny icon included along with a website, which is displayed in places like the browser's address bar, page tabs and bookmarks menu. Developers would like to have an API which allows them to retrieve the Favicon of a webpage, if it has been set, as well as get an update whenever the favicon has changed.
+
+# Description
+We propose a new Webview2 event which would allow developers to access the current Favicon of a page and be notified when the favicon changes. This means when a page first loads, it would raise the FaviconChanged event since before parsing the HTML document there is no known favicon. DOM or JavaScript may change the Favicon, causing the event to be raised again for the same document.
+
+# Examples
+## Win32 C++ Registering a listener for favicon changes
+```cpp
+CHECK_FAILURE(m_webView->add_FaviconChanged(
+    Callback<ICoreWebView2FaviconChangedEventHandler>(
+    [this](ICoreWebView2* sender, IUnknown* args) -> HRESULT {
+
+    wil::com_ptr<IStream> iconStream = SHCreateMemStream(nullptr, 0);
```

Comment:
This is not a common pattern. Usually GetIcon returns a stream that you can read from or (in this case) pass to GDI+. It doesn't usually require you to pass in a stream that it will write to.

The "give me a stream and I will write to it" pattern has a lot of open questions.

* Does the method rewind the stream to 0 before writing, or does it write starting at the stream's current position?
* Where is the stream pointer placed when the method returns?

It's not clear from the sample whether the method rewinds the stream before writing (because the sample gives the method an already-rewound stream).

The sample implies that on return, the stream pointer is set to the start of the icon data.

I think a better pattern would be for the implementation to create a new stream, load the pixels into it, rewind the stream, and then pass the stream to the completion handler.

```cpp
    sender->GetFavicon(COREWEBVIEW2_FAVICON_IMAGE_FORMAT_PNG,
        Callback<ICoreWebView2ExperimentalGetFaviconCompletedHandler>(
            [this](HRESULT error_code, IStream* stream) -> HRESULT
...
```

Reply:
We were following the pattern from an existing CoreWebView2.CapturePreviewAsync (which itself was based on pattern from previous WebView1 method), but I agree with your points and a method like you propose should be much easier to figure out and use.

Reply:
When we ported CapturePreviewAsync to Windows Runtime, we renamed the method to CapturePreviewToStreamAsync to make it clearer that it fills in a stream rather than producing one.

Reply:
Changing GetFavicon to return a stream just seems much easier for the end dev - much less to figure out. Unless folks have reasons why taking a stream as a parameter is better. 

For large amounts of data you may want to let the end dev provide the stream so there's less copying of data if they need to write it directly to a file or shared memory or they have some other requirements on the stream. I don't think that's really an issue for favicons though.

Reply:
Updated both this and the .Net code.